compiler/dialects/acc.py

```diff
@@ -0,0 +1,163 @@
+from collections.abc import Iterable, Sequence
+
+from xdsl.dialects.builtin import ArrayAttr, StringAttr
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+    Operation,
+    ParametrizedAttribute,
+    SSAValue,
+    TypeAttribute,
+)
+from xdsl.irdl import (
+    AttrSizedOperandSegments,
+    IRDLOperation,
+    ParameterDef,
+    VerifyException,
+    irdl_attr_definition,
+    irdl_op_definition,
+    operand_def,
+    opt_operand_def,
+    prop_def,
+    result_def,
+    var_operand_def,
+)
+
+
+@irdl_attr_definition
+class TokenType(ParametrizedAttribute, TypeAttribute):
+    """
+    Async token type for launched accelerator requests.
+    """
+
+    name = "acc.token"
+
+
+@irdl_attr_definition
+class StateType(ParametrizedAttribute, TypeAttribute):
+    """
+    Used to trace an accelerators CSR state through def-use chain
+    """
+
+    name = "acc.state"
+
+    accelerator: ParameterDef[StringAttr]
+
+
+@irdl_op_definition
+class LaunchOp(IRDLOperation):
+    """
+    Launch an accelerator. This acts as a barrier for CSR values,
+    meaning CSRs can be safely modified after a launch op without
+    interfering with the Accelerator.
+    """
+
+    name = "acc.launch"
+
+    accelerator = prop_def(StringAttr)
+
+    token = result_def()
+
+    def verify_(self) -> None:
+        # that the token is used
+        if len(self.token.uses) != 1 or not isinstance(
+            next(iter(self.token.uses)).operation, AwaitOp
+        ):
+            raise VerifyException("Launch token must be used by exactly one await op")
+        # TODO: allow use in control flow
+
+
+@irdl_op_definition
+class AwaitOp(IRDLOperation):
+    name = "acc.await"
+
+    token = operand_def(TokenType)
+
+
+@irdl_op_definition
+class SetupOp(IRDLOperation):
+    name = "acc.setup"
+
+    values = var_operand_def(Attribute)  # TODO: make more precise?
+    """
+    The actual values used to set up the CSRs
+    """
+
+    in_state = opt_operand_def(StateType)
+    """
+    The state produced by a previous acc.setup
+    """
+
+    out_state = result_def(StateType)
+    """
+    The CSR state after the setup op modified it.
+    """
+
+    param_names = prop_def(ArrayAttr[StringAttr])
+    """
+    Maps the SSA values in `values` to accelerator parameter names
+    """
+
+    accelerator = prop_def(StringAttr)
+    """
+    Name of the accelerator this setup is for
+    """
+
+    irdl_options = [AttrSizedOperandSegments(as_property=True)]
+
+    def __init__(
+        self,
+        vals: Sequence[SSAValue],
+        param_names: Sequence[str],
+        accelerator: str | StringAttr,
+        in_state: SSAValue | Operation | None = None,
+    ):
+        if not isinstance(accelerator, StringAttr):
+            accelerator = StringAttr(accelerator)
+
```

Comment:
Adding support for existing StringAttr params can be nice
```suggestion
        param_names: Sequence[str | StringAttr],
        accelerator: str | StringAttr,
        in_state: SSAValue | Operation | None = None,
    ):
        if not isinstance(accelerator, StringAttr):
            accelerator = StringAttr(accelerator)
        
        param_names = [
            StringAttr(param_name) if not isinstance(param_name, StringAttr) 
            else param_name 
            for param_name in param_names
        ]
```

Reply:
Yeah, we had this before, but pyright can't infer the types. We can add it, but it's not needed so far, and the way we ended up doing it in xDSL was:

```py
param_names : Sequence[str] | Sequence[StringAttr]  # cannot be a mixed sequence

# ...
params = tuple(param_names) # `isa` doesn't support Sequence, so cast to tuple
if not isa(params, tuple[StringAttr, ...]):
    params = tuple(StringAttr(x) for x in params
```
But this is not really nice looking